src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs

```diff
@@ -0,0 +1,354 @@
+
+        // TODO: _launchContextFields contain a list of fields that we will transmit as part of launch context, should be configurable
+        private readonly string[] _launchContextFields = { "patient", "encounter", "practitioner", "need_patient_banner", "smart_style_url" };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AadProxyController" /> class.
+        /// </summary>
+        /// <param name="securityConfiguration">Security configuration parameters.</param>
+        /// <param name="httpClientFactory">HTTP Client Factory.</param>
+        /// <param name="logger">The logger.</param>
+        public AadProxyController(IOptions<SecurityConfiguration> securityConfiguration, IHttpClientFactory httpClientFactory, ILogger<SecurityConfiguration> logger)
+        {
+            EnsureArg.IsNotNull(securityConfiguration, nameof(securityConfiguration));
+
+            _securityConfiguration = securityConfiguration.Value;
+            _isAadV2 = new Uri(_securityConfiguration.Authentication.Authority).Segments.Contains("v2.0");
+            _logger = logger;
+            _httpClientFactory = httpClientFactory;
+
+            var openIdConfigurationUrl = $"{_securityConfiguration.Authentication.Authority}/.well-known/openid-configuration";
+
+            HttpResponseMessage openIdConfigurationResponse;
+            using (var httpClient = httpClientFactory.CreateClient())
+            {
+                try
+                {
+                    openIdConfigurationResponse = httpClient.GetAsync(new Uri(openIdConfigurationUrl)).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    logger.LogWarning(ex, $"There was an exception while attempting to read the OpenId Configuration from \"{openIdConfigurationUrl}\".");
+                    throw new OpenIdConfigurationException();
+                }
+            }
+
+            if (openIdConfigurationResponse.IsSuccessStatusCode)
+            {
+                var openIdConfiguration = JObject.Parse(openIdConfigurationResponse.Content.ReadAsStringAsync().GetAwaiter().GetResult());
+
+                try
+                {
+                    _aadTokenEndpoint = openIdConfiguration["token_endpoint"].Value<string>();
+                    _aadAuthorizeEndpoint = openIdConfiguration["authorization_endpoint"].Value<string>();
+                }
+                catch (ArgumentNullException ex)
+                {
+                    logger.LogError($"{ex.Message}, There was an error attempting to read the endpoints from \"{openIdConfigurationUrl}\".");
+                    throw new OpenIdConfigurationException();
+                }
+            }
+            else
+            {
+                throw new OpenIdConfigurationException();
+            }
+        }
+
+        /// <summary>
+        /// Proxies a request to the Azure AD authorize endpoint.
```

Comment:
Might be more accurate to say that it redirects to the AAD authorize endpoint with adjusted parameters.